lib/rototiller/task/params/command.rb

```diff
@@ -161,15 +160,37 @@ def run
         #   we may have to convert this to a class if it gets complex
         @result = Result.new
         @result.output = ''
-        # add ';' to command string as it is a metacharacter that forces open3
-        #   to send the command to the shell. This returns the correct
-        #   exit_code and stderr, etc when the command is not found
-        Open3.popen2e(self.to_str + ";"){|stdin, stdout_err, wait_thr|
-          stdout_err.each { |line| puts line
-                            @result.output << line }
-          @result.pid    = wait_thr.pid # pid of the started process.
-          @result.exit_code = wait_thr.value.exitstatus # Process::Status object returned.
-        }
+
+        read_pipe, write_pipe = IO.pipe
+        begin
+          @result.pid = Process.spawn(self.to_str, :out => write_pipe, :err => write_pipe)
+        rescue Errno::ENOENT => e
+          $stderr.puts e
+          @result.output << e.to_s
+          @result.exit_code = 127
+          raise
+        end
+        #create a thread that monitors the process and tells us when it is done
+        @exitstatus = :not_done
+        Thread.new do
+          Process.wait(@result.pid);
+          @exitstatus       = $?.exitstatus
+          @result.exit_code = @exitstatus
+          write_pipe.close
+        end
+
+        #FIXME: monitor for deadlock?
+        while @exitstatus == :not_done
+          # could probably optimize a bit here by checking the pipe before writing/sleeping?
+          sleep 0.001
+          begin
+            this_read = read_pipe.readline
```

Comment:
I think if you measure how long it takes, it would take (6600 * 0.001) seconds to complete since it waits 0.001 seconds to read the next line even if multiple lines have been written, whereas it should complete faster if you use readline(nil) instead of just readline.  It looks like readline takes the same sep parameter, so doesn't need to be gets.